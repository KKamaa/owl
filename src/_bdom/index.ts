import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
