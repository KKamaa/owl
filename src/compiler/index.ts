import { setupBlockDom } from "./helpers";
import { QWebCompiler, TemplateFunction } from "./qweb_compiler";
export { Template, TemplateFunction } from "./qweb_compiler";

setupBlockDom();

export function compileTemplate(template: string, name?: string): TemplateFunction {
  const compiler = new QWebCompiler(template, name);
  return compiler.compile();
}
