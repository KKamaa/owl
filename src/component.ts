import type { BNode } from "./b_node";

export class Component {
  static template: string = "";

  props: any;
  env: any;
  __owl__: BNode;

  constructor(props: any, env: any, node: BNode) {
    this.props = props;
    this.env = env;
    this.__owl__ = node;
  }
  get el(): HTMLElement | null {
    const node = this.__owl__;
    return node.bdom ? (node.bdom.el as any) : null;
  }

  setup() {}

  render(): Promise<void> {
    return this.__owl__.render();
  }
}
