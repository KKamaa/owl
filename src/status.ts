import type { Component } from "./component";

export const enum STATUS {
  NEW,
  WILLSTARTED, // willstart has been called
  CREATED, // after first render is complete
  MOUNTED, // is ready, and in DOM. It has a valid el
  UNMOUNTED, // has a valid el, but is no longer in DOM
  DESTROYED,
}

type STATUS_DESCR = "new" | "willstarted" | "created" | "mounted" | "unmounted" | "destroyed";

export function status(component: Component): STATUS_DESCR {
  switch (component.__owl__.status) {
    case STATUS.NEW:
      return "new";
    case STATUS.WILLSTARTED:
      return "willstarted";
    case STATUS.CREATED:
      return "created";
    case STATUS.MOUNTED:
      return "mounted";
    case STATUS.UNMOUNTED:
      return "unmounted";
    case STATUS.DESTROYED:
      return "destroyed";
  }
}
