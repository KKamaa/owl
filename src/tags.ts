// -----------------------------------------------------------------------------
//  Global templates
// -----------------------------------------------------------------------------

export const globalTemplates: { [key: string]: string } = {};

export function xml(strings: TemplateStringsArray, ...args: any[]) {
  const name = `__template__${xml.nextId++}`;
  const value = String.raw(strings, ...args);
  globalTemplates[name] = value;
  return name;
}

xml.nextId = 1;
