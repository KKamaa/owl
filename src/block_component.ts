import { Block } from "./bdom";
import { makeChildFiber, __internal__destroyed } from "./fibers";
import { OwlNode } from "./owl_node";
import { STATUS } from "./status";

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

export class BComponent extends Block {
  node: OwlNode;
  parentClass?: any;
  classTarget?: HTMLElement;

  /**
   *
   * @param name
   * @param props
   * @param key
   * @param owner the component in which the component was defined
   * @param parent the actual parent (may be different in case of slots)
   */
  constructor(name: string, props: any, key: string, owner: any, parent: any) {
    super();
    const parentNode: OwlNode = parent.__owl__;
    let node: OwlNode | undefined = parentNode.children[key];
    if (node && node.status < STATUS.MOUNTED) {
      node.destroy();
      // delete parentNode.children[key];
      node = undefined;
    }
    const parentFiber = parentNode.fiber!;
    if (node) {
      // update
      const fiber = makeChildFiber(node, parentFiber);
      if (node.willPatch.length) {
        parentFiber.root.willPatch.push(fiber);
      }
      if (node.patched.length) {
        parentFiber.root.patched.push(fiber);
      }
      node.updateAndRender(props, fiber);
    } else {
      // new component
      const components = owner.constructor.components;
      const C = components[name];
      node = new OwlNode(parentNode.app, C, props);
      parentNode.children[key] = node;
      const fiber = makeChildFiber(node, parentNode.fiber!);
      if (node.mounted.length) {
        parentFiber.root.mounted.push(fiber);
      }
      node.initiateRender(fiber);
    }
    this.node = node;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.node.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    const node = this.node;
    const bdom = node.fiber!.bdom!;
    node.bdom = bdom;
    bdom.mountBefore(anchor);
    if (this.parentClass) {
      this.parentClass = this.parentClass.trim().split(/\s+/);
      const el = this.firstChildNode();
      if (el instanceof HTMLElement) {
        this.addClass(el);
      }
    }
    node.status = STATUS.MOUNTED;
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  moveBefore(anchor: ChildNode) {
    this.node.bdom!.moveBefore(anchor);
  }

  addClass(el: HTMLElement) {
    this.classTarget = el;
    for (let cl of this.parentClass) {
      el.classList.add(cl);
    }
  }

  removeClass(el: HTMLElement) {
    for (let cl of this.parentClass) {
      el.classList.remove(cl);
    }
  }

  patch() {
    const node = this.node;

    node.bdom!.patch(node!.fiber!.bdom!);
    if (this.parentClass) {
      const el = this.firstChildNode();
      if (el !== this.classTarget) {
        if (el && this.classTarget) {
          this.removeClass(this.classTarget);
          this.addClass(el as any);
        } else if (el) {
          this.addClass(el as any);
        } else {
          this.removeClass(this.classTarget!);
          this.classTarget = undefined;
        }
      }
    }
    node.fiber!.appliedToDom = true;
    node.fiber = null;
  }

  beforeRemove() {
    visitRemovedNodes(this.node);
  }

  remove() {
    const bdom = this.node.bdom!;
    bdom.remove();
  }
}

function visitRemovedNodes(node: OwlNode) {
  if (node.status === STATUS.MOUNTED) {
    const component = node.component;
    for (let cb of node.willUnmount) {
      cb.call(component);
    }
  }
  for (let child of Object.values(node.children)) {
    visitRemovedNodes(child);
  }
  node.status = STATUS.DESTROYED;
  if (node.destroyed.length) {
    __internal__destroyed.push(node);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, owner: any, parent: any) {
    super(name, props, key, owner, parent);
    this.handlers = new Array(handlers);
  }
  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }
  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.node.component.el!;
      el.addEventListener(eventType, (ev: Event) => {
        const info = this.handlers![i];
        const [, callback] = info;
        callback(ev);
      });
    }
  }
}
