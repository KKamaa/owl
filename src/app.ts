import { Blocks as BaseBlocks } from "./bdom";
import { compileTemplate, Template } from "./compiler/index";
import { globalTemplates } from "./tags";
import { BComponent, BComponentH } from "./block_component";
import type { Component } from "./component";
import { OwlNode } from "./owl_node";
import { Scheduler } from "./scheduler";
import { UTILS } from "./template_utils";

const Blocks = {
  ...BaseBlocks,
  BComponent,
  BComponentH,
};

// -----------------------------------------------------------------------------
//  TemplateSet
// -----------------------------------------------------------------------------

export class TemplateSet {
  rawTemplates: { [name: string]: string } = Object.create(globalTemplates);
  templates: { [name: string]: Template } = {};
  utils: typeof UTILS;

  constructor() {
    const call = (subTemplate: string, ctx: any, parent: any) => {
      const template = this.getTemplate(subTemplate);
      return template(ctx, parent);
    };

    this.utils = Object.assign({}, UTILS, { call });
  }

  addTemplate(name: string, template: string, options: { allowDuplicate?: boolean } = {}) {
    if (name in this.rawTemplates && !options.allowDuplicate) {
      throw new Error(`Template ${name} already defined`);
    }
    this.rawTemplates[name] = template;
  }

  getTemplate(name: string): Template {
    if (!(name in this.templates)) {
      const rawTemplate = this.rawTemplates[name];
      if (rawTemplate === undefined) {
        throw new Error(`Missing template: "${name}"`);
      }
      const templateFn = compileTemplate(rawTemplate, name);
      const template = templateFn(Blocks, this.utils);
      this.templates[name] = template;
    }
    return this.templates[name];
  }
}

export class App<T extends typeof Component = any> extends TemplateSet {
  Root: T;
  props: any;
  env: any = {};
  scheduler = new Scheduler(window.requestAnimationFrame.bind(window));
  root: OwlNode | null = null;
  dev: boolean = true;
  
  constructor(Root: T, props?: any) {
    super();
    this.Root = Root;
    this.props = props;
  }

  configure(params: any) {
    if (params.env) {
      this.env = params.env;
    }
  }

  mount(target: HTMLElement): Promise<InstanceType<T>> {
    if (!(target instanceof HTMLElement)) {
      throw new Error("Cannot mount component: the target is not a valid DOM element");
    }
    if (!document.body.contains(target)) {
      throw new Error("Cannot mount a component on a detached dom node");
    }
    const node = new OwlNode(this, this.Root, this.props);
    this.root = node;
    return node.mount(target);
  }

  destroy() {
    if (this.root) {
      this.root.destroy();
    }
  }
}
