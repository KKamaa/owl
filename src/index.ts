import { App } from "./app";
import { Component } from "./component";

export { App, Component };

export async function mount<T extends typeof Component>(
  C: T,
  target: HTMLElement
): Promise<InstanceType<T>> {
  const app = new App(C);
  return app.mount(target);
}

export { xml } from "./tags";
export { useState } from "./reactivity";
export { useRef } from "./refs";
export { EventBus } from "./event_bus";

export {
  onWillStart,
  onMounted,
  onBeforeUnmount,
  onWillUpdateProps,
  onBeforePatch,
  onPatched,
} from "./lifecycle_hooks";

export const __info__ = {};
