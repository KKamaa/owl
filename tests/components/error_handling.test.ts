import { Component, mount } from "../../src";
import { status } from "../../src/status";
import { xml } from "../../src/tags";
import { makeTestFixture, snapshotEverything } from "../helpers";

let fixture: HTMLElement;

snapshotEverything();

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("basics", () => {
  test("no component catching error lead to full app destruction", async () => {
    // expect.assertions(6);
    // const handler = jest.fn();
    // const consoleError = console.error;
    // console.error = jest.fn();

    class ErrorComponent extends Component {
      static template = xml`<div>hey<t t-esc="props.flag and state.this.will.crash"/></div>`;
    }

    class Parent extends Component {
      static template = xml`<div><ErrorComponent flag="state.flag"/></div>`;
      static components = { ErrorComponent };
      state = { flag: false };
    }
    const parent = await mount(Parent, fixture);
    expect(fixture.innerHTML).toBe("<div><div>heyfalse</div></div>");
    parent.state.flag = true;

    let error;
    try {
      await parent.render();
    } catch (e) {
      error = e;
    }
    expect(fixture.innerHTML).toBe("");
    expect(status(parent)).toBe("destroyed");
    expect(error).toBeDefined();
    expect(error.message).toBe("cannot");

    // expect(console.error).toBeCalledTimes(0);
    // console.error = consoleError;
    // expect(handler).toBeCalledTimes(1);
  });
});
