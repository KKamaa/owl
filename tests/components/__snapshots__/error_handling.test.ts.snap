// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basics no component catching error lead to full app destruction 1`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div>hey<owl-text/></div>\`);
    data = new Array(1);
    update() {
      this.el.firstChild.nextSibling.textContent = toString(this.data[0]);
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.data[0] = ctx['props'].flag&&ctx['state'].this.will.crash;
    return b1;
  }
}"
`;

exports[`basics no component catching error lead to full app destruction 2`] = `
"function anonymous(Blocks, utils
) {
  let {BCollection, BComponent, BComponentH, BHtml, BMulti, BNode, BStatic, BText, BDispatch} = Blocks;
  let {elem, toString, withDefault, call, getTemplate, zero, callSlot, capture} = utils;
  
  class Block1 extends BNode {
    static el = elem(\`<div><owl-anchor/></div>\`);
    children = new Array(1);
    anchors = new Array(1);
    build() {
      this.anchors[0] = this.el.firstChild;
    }
  }
  
  return function template(ctx, node = ctx.__owl__, key = \\"\\") {
    const b1 = new Block1();
    b1.children[0] = new BComponent(\`ErrorComponent\`, {flag: ctx['state'].flag}, key + \`__1\`, ctx, node);
    return b1;
  }
}"
`;
